src/api.ts

```diff
@@ -260,6 +281,92 @@ function createApi({
   }
 
   const api: Api = {
+    async importMBTiles(filePath: string) {
+      const filePathWithExtension =
+        path.extname(filePath) === '.mbtiles' ? filePath : filePath + '.mbtiles'
+
+      let mbTilesDb: DatabaseInstance
+
+      try {
+        mbTilesDb = new Database(filePathWithExtension, {
+          // Ideally would set `readOnly` to `true` here but causes `fileMustExist` to be ignored:
+          // https://github.com/JoshuaWise/better-sqlite3/blob/230ea65ed0d7566e32d41c3d13a90fb32ccdbee6/docs/api.md#new-databasepath-options
+          fileMustExist: true,
+        })
+      } catch (_err) {
+        throw new MBTilesImportTargetMissingError(filePath)
+      }
+
+      const tilejson = mbTilesToTileJSON(mbTilesDb)
+
+      // TODO: Should this be handled in extractMBTilesMetadata?
+      if (!(tilejson.format && isValidMBTilesFormat(tilejson.format))) {
+        throw new UnsupportedMBTilesFormatError()
+      }
+
+      if (!validateTileJSON(tilejson)) {
+        mbTilesDb.close()
+        throw new MBTilesInvalidMetadataError()
+      }
+
+      const insertTileData = db.prepare<{
+        data: Buffer
+        tileHash: string
+        tilesetId: string
+      }>(
+        'INSERT INTO TileData (tileHash, data, tilesetId) VALUES (:tileHash, :data, :tilesetId)'
+      )
+
+      const insertTile = db.prepare<{
+        quadKey: string
+        tileHash: string
+        tilesetId: string
+      }>(
+        'INSERT INTO Tile (quadKey, tileHash, tilesetId) VALUES (:quadKey, :tileHash, :tilesetId)'
+      )
+
+      const tilesetId = getTilesetId(tilejson)
+
+      // TODO: Look into whether we need to have a transaction for this
+      // Concerns regarding memory due to tile data size
+      const tilesImportTransaction = db.transaction(async () => {
```

Comment:
@gmaclennan highlighting this since this is something you pointed out before. haven't looked into it just yet

Reply:
A [transaction cannot be an async function](https://github.com/JoshuaWise/better-sqlite3/blob/HEAD/docs/api.md#caveats)

Reply:
I don't think we should use transactions for mbtiles inserts. The purpose of a transaction is to enable rollback if the transaction should fail before it completes. As I understand it, this is done by writing a [rollback file](https://www.sqlite.org/tempfiles.html) and deleting it at the end of the transaction. Since these tile import transactions can include large amounts of data, this could require a user to have twice as much disk space free than what is needed for the tiles in order to import.
I don't see there being an issue if we cannot rollback a tile import, e.g. if the app crashes before all the tiles are imported. I think we should create the tileset first, so that exists in case of failure, then add the tiles, but without a transaction.

Reply:
makes sense. will go with the suggestion you have in your other comment 👍